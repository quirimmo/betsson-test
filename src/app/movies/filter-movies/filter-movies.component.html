<form>
	<div class="row search-movie-container">
		<mat-form-field class="col-md-6 col-12 w-100">
			<input
				id="search-movie-name"
				[formControl]="nameText"
				matInput
				placeholder="Type the name of the movie..."
			/>
		</mat-form-field>
		<mat-form-field class="col-md-6 col-12 w-100">
			<mat-select
				id="seach-movie-genres"
				placeholder="Select the genres..."
				[formControl]="genresSelect"
				multiple
			>
				<mat-option *ngFor="let genre of genres" [value]="genre">{{
					genre
				}}</mat-option>
			</mat-select>
		</mat-form-field>
	</div>
</form>
