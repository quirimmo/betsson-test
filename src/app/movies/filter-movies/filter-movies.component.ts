import { Component, Input, OnInit, OnDestroy } from '@angular/core';
import { FormControl } from '@angular/forms';
import { Subscription } from 'rxjs';
import { ActivatedRoute, Router } from '@angular/router';
import { MoviesRouterService } from '../movies-router.service';

@Component({
	selector: 'filter-movies',
	inputs: ['genres'],
	styleUrls: ['./filter-movies.component.scss'],
	templateUrl: './filter-movies.component.html'
})
export class FilterMoviesComponent implements OnInit, OnDestroy {
	@Input()
	genres: string[];
	genresSelect = new FormControl();
	nameText = new FormControl();
	nameTextChangesSubscriber: Subscription;
	genresChangesSubscriber: Subscription;

	constructor(
		private moviesRouterService: MoviesRouterService,
		private route: ActivatedRoute,
		private router: Router
	) {}

	ngOnInit(): void {
		const searchName: string =
			this.moviesRouterService.getSearchNameParam() || '';
		const searchGenres: string[] =
			this.moviesRouterService.getSearchGenresParam() || this.genres;

		this.nameTextChangesSubscriber = this.nameText.valueChanges.subscribe(
			this.onChangeSearchName.bind(this)
		);
		this.genresChangesSubscriber = this.genresSelect.valueChanges.subscribe(
			this.onChangeSearchGenres.bind(this)
		);

		this.nameText.setValue(searchName);
		this.genresSelect.setValue(searchGenres);
	}

	ngOnDestroy(): void {
		this.nameTextChangesSubscriber.unsubscribe();
		this.genresChangesSubscriber.unsubscribe();
	}

	public onChangeSearchName(): void {
		this.moviesRouterService.appendSearchNameParam(this.nameText.value);
	}

	public onChangeSearchGenres(): void {
		this.moviesRouterService.appendSearchGenresParam(this.genresSelect.value);
	}
}
