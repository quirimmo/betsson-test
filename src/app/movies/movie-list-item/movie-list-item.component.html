<section class="movie-item-card">
	<div class="movie-item-card-title">{{ movie.name }}</div>
	<div>{{ movie.rate }}</div>
	<img mat-card-image [src]="'assets/' + movie.img" alt="{{ movie.name }}" />
	<div>{{ movie.genres.join(',') }}</div>
</section>
