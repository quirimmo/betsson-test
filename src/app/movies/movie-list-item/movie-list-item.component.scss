@import './../../../colors';
.movie-item-card {
	text-align: center;
	position: relative;
	margin-bottom: 30px;
	color: $blue;
}

.movie-item-card-title {
	min-height: 50px;
	font-size: 17pt;
}

.movie-item-card-rate {
	background-color: $yellow;
	color: $black;
	display: inline-block;
	top: -15px;
	padding: 2px 10px;
	position: relative;
	max-width: fit-content;
	margin: 0 auto;
}
