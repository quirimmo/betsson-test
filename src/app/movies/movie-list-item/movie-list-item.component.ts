import { Component, Input, OnInit } from '@angular/core';
import Movie from '../movie.model';

@Component({
	selector: 'movie-list-item',
	inputs: ['movie'],
	styleUrls: ['./movie-list-item.component.scss'],
	templateUrl: './movie-list-item.component.html'
})
export class MovieListItemComponent {
	@Input()
	movie: Movie;

	constructor() {}
}
