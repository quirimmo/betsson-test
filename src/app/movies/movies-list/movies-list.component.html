<section>
	<section>
		<filter-movies #filterMovies [genres]="genres"></filter-movies>
	</section>
	<section class="row">
		<div
			*ngIf="
				isNoMoviesMessageVisible(
					filterMovies.nameText,
					filterMovies.genresSelect
				)
			"
			class="col-12 justify-content-center"
			[@slideInLeftOnEnter]
			[@slideOutRightOnLeave]
		>
			<p class="no-movies-list-message">
				It looks like there are not movies corresponding to your search...
				<br />
				Try changing your filters!
			</p>
		</div>
		<div
			*ngIf="
				!isMoviesListEmpty(
					filterMovies.nameText.value,
					filterMovies.genresSelect.value
				)
			"
			class="col-12"
		>
			<div class="row">
				<section class="col-12 mb-5">
					<sort-movies #sortMovies></sort-movies>
				</section>
				<article
					class="col-xl-2 col-lg-3 col-md-4 col-sm-6 col-12 movie-list-item-wrapper"
					*ngFor="
						let movie of (getFilteredMovies(
							filterMovies.nameText.value,
							filterMovies.genresSelect.value
						) | sortMovies: sortMovies.sortBySelect.value);
						trackBy: trackById
					"
					[@bounceInUpOnEnter]
					[@bounceOutOnLeave]
				>
					<a [routerLink]="['/movie', movie.id]">
						<movie-list-item [movie]="movie"></movie-list-item>
					</a>
				</article>
			</div>
		</div>
	</section>
</section>
