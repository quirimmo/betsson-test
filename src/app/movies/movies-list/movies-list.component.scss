.search-movie-container {
  margin: 20px 0px;
  text-align: center;
}

.movie-list-item-wrapper {
  opacity: 0.75;
  -webkit-transition: all 0.35s;
  transition: all 0.35s;
}

.movie-list-item-wrapper:hover {
  opacity: 1;
}

.movie-list-item-wrapper a:hover {
  text-decoration: none;
}

.no-movies-list-message {
  text-align: justify;
  background: #ead430;
  color: #000405;
  font-size: 13pt;
  padding: 20px;
  border-radius: 5px;
}
