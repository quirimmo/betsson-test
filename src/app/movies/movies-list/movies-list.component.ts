import { Component, Input, OnInit } from '@angular/core';
import Movie from '../movie.model';

@Component({
	selector: 'movies-list',
	inputs: ['movies'],
	styleUrls: ['./movies-list.component.scss'],
	templateUrl: './movies-list.component.html'
})
export class MoviesListComponent implements OnInit {
	@Input()
	movies: Movie[];

	constructor() {}

	ngOnInit(): void {
		console.log('movies in the comp:', this.movies);
	}
}
