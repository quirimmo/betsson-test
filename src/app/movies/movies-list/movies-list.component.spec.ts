import { TestBed, async } from '@angular/core/testing';
import Movie from 'src/app/movies/movie.model';
import { FilterMoviesPipe } from 'src/app/movies/pipes/filter-movies.pipe';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { MoviesListComponent } from 'src/app/movies/movies-list/movies-list.component';
import { FilterMoviesComponent } from '../filter-movies/filter-movies.component';
import { RouterTestingModule } from '@angular/router/testing';
import GenreType from '../genre.model';
import { MovieListItemComponent } from '../movie-list-item/movie-list-item.component';

const movie = new Movie(
	1,
	'key',
	'name',
	'description',
	[GenreType.Action],
	'8.4',
	'1h 23m',
	'img.jpg'
);
const movies = [movie];
const mockTransform = jest.fn().mockReturnValue(movies);
const mockFilterMoviesPipe = {
	provide: FilterMoviesPipe,
	useValue: {
		transform: mockTransform
	}
};
const providers = [mockFilterMoviesPipe];
const imports = [RouterTestingModule, BrowserAnimationsModule];
const declarations = [
	FilterMoviesComponent,
	MovieListItemComponent,
	MoviesListComponent
];

const MockFilterMoviesComponent = {
	set: {
		inputs: ['genres'],
		template: '<span>filter movies</span>'
	}
};
const MockMovieListItemComponent = {
	set: {
		inputs: ['movie'],
		template: '<span>movie list item</span>'
	}
};

describe('MoviesList Component', () => {
	let componentFixture, componentInstance;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			declarations,
			providers,
			imports
		})
			.overrideComponent(FilterMoviesComponent, MockFilterMoviesComponent)
			.overrideComponent(MovieListItemComponent, MockMovieListItemComponent)
			.compileComponents();
		componentFixture = TestBed.createComponent(MoviesListComponent);
		componentInstance = componentFixture.debugElement.componentInstance;
		componentInstance.movies = movies;
	}));

	it('should create the component', () => {
		expect(componentFixture).toBeTruthy();
	});

	it('should match the snapshot', () => {
		expect(componentFixture).toMatchSnapshot();
	});

	describe('trackById', () => {
		it('should return the id of the movie', () => {
			expect(componentInstance.trackById(0, movie)).toEqual(1);
		});
	});

	describe('getFilteredMovies', () => {
		it('should call the transform method of the FilterMoviesPipe', () => {
			componentInstance.getFilteredMovies('', []);
			expect(mockTransform).toHaveBeenCalledWith(movies, '', []);
		});
	});
});
