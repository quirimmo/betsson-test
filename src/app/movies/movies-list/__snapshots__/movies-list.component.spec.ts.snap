// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList Component should match the snapshot 1`] = `
<movies-list
  filterMoviesPipe={[Function Object]}
  genres={[Function Array]}
  movies={[Function Array]}
>
  <section
    class="ng-tns-c1-0"
  >
    <section
      class="mb-4"
    >
      <filter-movies
        class="ng-tns-c1-0"
      >
        <span>
          filter movies
        </span>
      </filter-movies>
    </section>
    <section
      class="row"
    >
      
      
    </section>
  </section>
</movies-list>
`;
